CMakeLists.txt

```diff
@@ -624,30 +622,44 @@ if(AVIF_BUILD_APPS OR (AVIF_BUILD_TESTS AND (AVIF_ENABLE_FUZZTEST OR AVIF_ENABLE
     endif()
 
     add_library(
-        avif_apps STATIC apps/shared/avifexif.c apps/shared/avifjpeg.c apps/shared/avifpng.c apps/shared/avifutil.c
-                         apps/shared/iccmaker.c apps/shared/y4m.c third_party/iccjpeg/iccjpeg.c
+        avif_apps_obj OBJECT $<TARGET_OBJECTS:avif_obj> apps/shared/avifexif.c apps/shared/avifjpeg.c apps/shared/avifpng.c
+                             apps/shared/avifutil.c apps/shared/iccmaker.c apps/shared/y4m.c third_party/iccjpeg/iccjpeg.c
     )
-    target_link_libraries(avif_apps avif PNG::PNG ZLIB::ZLIB JPEG::JPEG)
+    set_property(TARGET avif_apps_obj PROPERTY POSITION_INDEPENDENT_CODE ON)
+    target_link_libraries(avif_apps_obj avif_obj PNG::PNG ZLIB::ZLIB JPEG::JPEG)
```

Comment:
IMPORTANT: I think this line causes the .c files in avif_apps_obj, such as avifjpeg.c, to be compiled with -DAVIF_DLL when -DBUILD_SHARED_LIBS=ON is specified.

Then, avifjpeg.c.o, compiled with -DAVIF_DLL, is added to both libavif_apps.a and libavif_apps_internal.a.

But the object files in libavif_apps_internal.a should be compiled with -DAVIF_USING_STATIC_LIBS, not -DAVIF_DLL.

Reply:
Well, that is already the case:
https://github.com/AOMediaCodec/libavif/blob/15fa36c4c1b1beb409ec7ef44e41fadf8a38b944/CMakeLists.txt#L524
and https://github.com/AOMediaCodec/libavif/blob/15fa36c4c1b1beb409ec7ef44e41fadf8a38b944/CMakeLists.txt#L630

It does not matter that much as avif_internal is not meant to be distributed. This only affects which symbols are visible and it has to be used at compile time: we would definitely need to compile the lib twice then.

Reply:
Thanks for the explanation. On Windows, when `-DBUILD_SHARED_LIBS=ON` is specified, if we compile avifjpeg.c with `-DAVIF_DLL,` then avifjpeg.c will see the `AVIF_API` macro defined as `__declspec(dllimport)`, and I believe a call to `avifImageAllocatePlanes` will be recorded as a reference to the symbol `__imp_avifImageAllocatePlanes` in the object file. If so, that reference can only be resolved by linking with the avif DLL import library. If we link the object file with only the avif_internal static library, then I believe we will get an unresolved symbol linker error for `__imp_avifImageAllocatePlanes`.

In the current code, since we either link with the avif DLL import library or link with both the avif DLL import library and the avif_internal static library, we will not see this unresolved symbol linker error. But in the previous version of this pull request, we would link the tests with only the avif_internal static library. Then the tests that use avif_apps would get the unresolved symbol linker error, I think. (I haven't verified this on my Windows laptop though.)

Reply:
I just tested this pull request on Windows. I verified that by compiling avifjpeg.c with `-DAVIF_DLL`, we recorded references to symbols such as `__imp_avifRGBImageSetDefaults` in the object file (avifjpeg.c.obj). But I was surprised to find that the static library avif_internal.lib can also satisfy the references to the `__imp_*` symbols. I don't know why.

So I think it is fine that files in avif_apps_obj are compiled only once, with `-DAVIF_DLL`.